Fix clearIfComposited for indexed color write masks

lexaknyazev opened the pull request:
#### d26dacdc2fc03f111cf24ad1e8fe116a4bdaf197
<pre>
Fix clearIfComposited for indexed color write masks
<a href="https://bugs.webkit.org/show_bug.cgi?id=241511">https://bugs.webkit.org/show_bug.cgi?id=241511</a>

Patch by Alexey Knyazev &lt;3479527+lexaknyazev@users.noreply.github.com &gt; on 2022-06-13
Reviewed by Kimmo Kinnunen.

Implicit clears must correctly restore color write mask state.

* Source/WebCore/html/canvas/OESDrawBuffersIndexed.cpp:
(WebCore::OESDrawBuffersIndexed::colorMaskiOES):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::clearIfComposited):
(WebCore::WebGLRenderingContextBase::restoreStateAfterClear):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:

Canonical link: <a href="https://commits.webkit.org/251490@main">https://commits.webkit.org/251490@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@295485">https://svn.webkit.org/repository/webkit/trunk@295485</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

kkinnunen-apple commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -1361,7 +1361,10 @@ bool WebGLRenderingContextBase::clearIfComposited(WebGLRenderingContextBase::Cle
     } else
         m_context->clearColor(0, 0, 0, 0);
-    m_context->colorMask(true, true, true, true);
+    if (m_oesDrawBuffersIndexed)
```
Is this hunk needed ?
I'm thinking the case 

set the color mask
enable the extension

vs

enable the extension
set the color mask


So these must end up in the same state inside ANGLE.
This should mean that  these must be interchangeable:

   colorMaskiOES(0, ...)
vs
   colorMask(...);

If they're interchangeable, then this hunk shouldn't be needed?

lexaknyazev replied:
This line checks if the extension was enabled by an application. If the extension was not enabled, ANGLE would reject `colorMaskiOES` calls.

kkinnunen-apple commented on Source/WebCore/html/canvas/OESDrawBuffersIndexed.cpp:
```diff
@@ -105,6 +105,14 @@ void OESDrawBuffersIndexed::colorMaskiOES(GCGLuint buf, GCGLboolean red, GCGLboo
     if (!m_context || m_context->isContextLost())
         return;
 
```
would this be more obvious as
```
 if (!buf) {
   m_context->colorMask(...);
   return;
 }
```
(if the calls are interchangeable, that is)

lexaknyazev replied:
`colorMask(...)` sets the mask for all draw buffers, while `colorMaskiOES(0, ...)` affects only one.

zdobersek committed:
[GBM] Overzealous attaching of GL texture as color attachment to the default framebuffer
https://bugs.webkit.org/show_bug.cgi?id=241462

Patch by Žan Doberšek <zdobersek@igalia.com> on 2022-06-13
Reviewed by Alejandro G. Castro.

In GraphicsContextGLGBM::prepareTexture(), there's no need to repeatedly
reattach the texture objec to the default framebuffer. With GraphicsContextGLGBM
this is only done when (re)sizing the framebuffer. The attached texture later
isn't switched out for some other texture, it remains in place. What does change
is the backing DMABuf-based EGLImage, which is updated upon the next
make-context-current request.

This means framebuffer operations here, after any required multisampling is
resolved, are not required. Only thing that remains is the GL_Flush() call.

* Source/WebCore/platform/graphics/gbm/GraphicsContextGLGBM.cpp:
(WebCore::GraphicsContextGLGBM::prepareTexture):

Canonical link: https://commits.webkit.org/251489@main
git-svn-id: https://svn.webkit.org/repository/webkit/trunk@295484 268f45cc-cd09-0410-ab3c-d52691b4dbfc

lexaknyazev committed:
Fix clearIfComposited for indexed color write masks
https://bugs.webkit.org/show_bug.cgi?id=241511

Patch by Alexey Knyazev <3479527+lexaknyazev@users.noreply.github.com> on 2022-06-13
Reviewed by Kimmo Kinnunen.

Implicit clears must correctly restore color write mask state.

* Source/WebCore/html/canvas/OESDrawBuffersIndexed.cpp:
(WebCore::OESDrawBuffersIndexed::colorMaskiOES):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::clearIfComposited):
(WebCore::WebGLRenderingContextBase::restoreStateAfterClear):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:

Canonical link: https://commits.webkit.org/251490@main
git-svn-id: https://svn.webkit.org/repository/webkit/trunk@295485 268f45cc-cd09-0410-ab3c-d52691b4dbfc
